add method for String for an IdentityMapping

cdwensley opened the pull request:
# Description
This PR adds a method for String for an IdentityMapping, and so attempts to deal with issue #3748.

## Text for release notes 

If this pull request should be mentioned in the release notes, 
please provide a short description matching the style of the GAP
[Changes manual](https://www.gap-system.org/Manuals/doc/changes/chap0.html).

## Further details
The method has been added at the end of file mapping.gi, but that may not be the correct place? 
It would be easy to add a test, if it was clear where to add it. 

# Checklist for pull request reviewers

- [ ] proper formatting

If your code contains kernel C code, run `clang-format` on it; the 
simplest way is to use `git clang-format`, e.g. like this (don't
forget to commit the resulting changes):

    git clang-format $(git merge-base HEAD master)

- [ ] usage of relevant labels

   1. either `release notes: not needed` or `release notes: to be added`
   2. at least one of the labels `bug` or `enhancement` or `new feature`
   3. for changes meant to be backported to `stable-4.X` add the `backport-to-4.X` label
   4. consider adding any of the labels `build system`, `documentation`, `kernel`, `library`, `tests`

- [ ] runnable tests
- [ ] lines changed in commits are sufficiently covered by the tests
- [ ] adequate pull request title
- [ ] well formulated text for release notes
- [ ] relevant documentation updates
- [ ] sensible comments in the code

fingolfin commented on lib/mapping.gi:
```diff
@@ -1597,3 +1597,14 @@ function(f,t)
+    str := Concatenation( "IdentityMapping( ", str, " )" ); 
+    ConvertToStringRep( str );
+    return str;
+end );
```
I think it would be better to put this near the `ViewObj` and `PrintObj` methods, and also match their filters and coding style. That would be in `lib/mapprep.gi`, roughly around line 1343, and it also involves dynamic rank adjustment.

cdwensley replied:
Now moved to mapprep.gi as suggested, added the rank adjustment, and simplified the code.

fingolfin commented on lib/mapping.gi:
```diff
@@ -1597,3 +1597,14 @@ function(f,t)
+    local str; 
+    str := String( Source( id ) ); 
+    str := Concatenation( "IdentityMapping( ", str, " )" ); 
+    ConvertToStringRep( str );
```
That shouldn't be needed, `Concatenation` for three inputs in string rep (which ought to be the case here) gives a string rep. Am I missing something?

cdwensley replied:
Removed ConvertToStringRep - it was just copied from some other methods.

ChrisJefferson replied:
This isn't a requirement, just some GAP advice.

You could write `return StringFormatted( "IdentityMapping( {} )", Source( id ) );`. For this case, there isn't really much difference between Concatenation and StringFormatted, but StringFormatted can get more useful for longer strings.

cdwensley replied:
I don't understand the syntax, but have made the change.

ChrisJefferson commented:
Can I make one final request (sorry, these should have come in all at once!) -- add a Test (you can add it to an existing test), just so we make sure it works, and doesn't break in future!

cdwensley commented:
I had a look yesterday in the test folders but could not find an appropriate file. 
Would it be OK to add something to testinstall/mapping.tst? 
There are a number of lines there involving DirectProductElements.

cdwensley commented:
Test added - is one instance sufficient?

fingolfin commented on tst/testinstall/mapping.tst:
```diff
@@ -403,5 +403,16 @@ true
+[ DirectProductElement( [ [ (1,2), (3,4) ] -> [ (3,4), (1,2) ], 
+      IdentityMapping( Group( [ (1,2), (3,4) ] ) ) ] ), 
+  DirectProductElement( [ IdentityMapping( Group( [ (1,2), (3,4) ] ) ), 
+      [ (1,2), (3,4) ] -> [ (3,4), (1,2) ] ] ) ]
```
That's good, but I think a much more direct test would be better (though this one can be left in as well). I.e., just something like this:
```
gap> String(IdentityMapping(SymmetricGroup(3)));
"IdentityMapping( Sym( [ 1 .. 3 ] ) )"
```
(That said, I am also fine with merging this PR as-is).

fingolfin commented on lib/mapprep.gi:
```diff
@@ -1366,6 +1367,17 @@ InstallMethod( PrintObj,
+    # the rank of IsGeneralMapping and IsOne happens to be increased a lot
+    {} -> SUM_FLAGS - RankFilter( IsGeneralMapping and IsOne ),
+  function ( id ) 
+    return StringFormatted( "IdentityMapping( {} )", Source(id) ); 
```
I guess @ChrisJefferson and me disagree here what is best; personally, I would not use `StringFormatted` here, as fewer people are familiar with it, so I think the construction with `Concatenation` is more maintainable. However, the code is fine either way, and in the end it's a matter of taste. So, no need to change this again, I just wanted to get it off my chest (being one of these blunt germans without manners, please forgive me :-)

fingolfin commented:
@cdwensley this looks good to me, thank you!

Would you mind squashing this PR down into a single commit? Alternatively, I can "squash and merge" it (it's just a button press, but the resulting commit will have me as author, I think, which I find a bit silly given that you did all the work). Anyway, whichever way you prefer is fine by me.

cdwensley commented:
This is an ideal opportunity for me to learn how to squash.  I just know that one starts with "git rebase -i", but I have never tried it.  I'll get someone to supervise me via Slack tomorrow morning.

fingolfin commented:
Sure thing. Basically, you can do `git rebase -i master`; this will enter an editor; in there, change all but the first appearance of the word `pick` at a line start to `squash` (or just `s` is enough); the text you'll be shown in the editor actually contains a summary of instructions at the end.

And if it fails, you can use `git rebase --abort` to get back to where you started.

cdwensley commented:
Thanks to lots of help on Slack I have now installed Atom and 'git rebase -i master' now works, providing me with a second atom file to which I added a commit message.  But nothing appears to have affected this PR page, so what's next to do? 
```
ChristoleysiMac:gapdev christopherwensley$ git config --global core.editor "atom --wait"
ChristoleysiMac:gapdev christopherwensley$ git rebase -i master
[detached HEAD 2d31c72e5] add method for String for an IdentityMapping
 Date: Wed Nov 20 15:13:34 2019 +0000
 2 files changed, 26 insertions(+), 1 deletion(-)
Successfully rebased and updated refs/heads/string.
```

ChrisJefferson commented:
Try typing `git rebase -i upstream/master` (replace upstream with whatever you use to refer to the github.com/gap-system branch of GAP). If you don't know if you have that, you can use `git remote -v`, post the output of that here (or on slack).

fingolfin commented:
You need to push, and indeed, "force push", via `git push -f`

fingolfin commented:
(Before that, perhaps check that `git log` outputs what you'd expect)

cdwensley commented:
Sorry - only just found the last few hints.  Am I doing this from the string branch, or from master? 
As the following log shows, I do not know what I am doing. 
```
ChristoleysiMac:gapdev christopherwensley$ git log
commit 2d31c72e5c6aa42b8a96253eaa98c5774b825821 (HEAD -> string)
Author: Chris Wensley <cwensley1234@btinternet.com>
Date:   Wed Nov 20 15:13:34 2019 +0000

    add method for String for an IdentityMapping
    
    moved the String method from mapping.gi to mapprep.gi
    
    using StringFormatted in place of Concatenation
    
    added a test at the end of tst/testinstall/mapping.tst
    
    apparently it is necessary to declare new variables
    
    and added the simpler test suggested by Max
    
    "squashing 6 commits for adding String method for IdentityMapping"

ChristoleysiMac:gapdev christopherwensley$ git remote -v
origin	https://github.com/cdwensley/gap (fetch)
origin	https://github.com/cdwensley/gap (push)
upstream	https://github.com/gap-system/gap (fetch)
upstream	https://github.com/gap-system/gap (push)
ChristoleysiMac:gapdev christopherwensley$ 
ChristoleysiMac:gapdev christopherwensley$ git push -f
fatal: The current branch string has no upstream branch.
To push the current branch and set the remote as upstream, use

    git push --set-upstream origin string

ChristoleysiMac:gapdev christopherwensley$ git rebase -i origin/master
Warning: the command isn't recognized in the following line:
 - rebase 2d31c72e5 add method for String for an IdentityMapping

You can fix this with 'git rebase --edit-todo' and then run 'git rebase --continue'.
Or you can abort the rebase with 'git rebase --abort'.
ChristoleysiMac:gapdev christopherwensley$ git rebase --abort
```

fingolfin commented:
Several remarks:
- I don't think the `git rebase -i origin/master` is useful here (and I am not quite sure why @ChrisJefferson suggested it in the first place); just do a `git rebase --abort` to get back to a normal state
- the `git log` shows that you didn't edit the commit message during the rebase, so it now just consists of the 6 old commit messages glued together. You may want to change that, e.g. by doing `git commit --amend`. That's the quickest way (although one could also do it via an interactive rebase)
- for details on both that and on interactive rebase, please consider reading https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History
- the `push` failed because your `string` branch has no upstream set. So how do you normally push changes to GitHub? Do exactly that, but add the `-f` / `--force` option
- as the text printed by `git push` indicates, one can instruct `git` to "remember" where to push a branch by default, via the `--set-upstream` / `-u` option. That's merely a convenience, but quite useful if one edits a pull request a lot.

cdwensley commented:
Thanks: so 'git push -f origin string' has caused something to happen, and tests have restarted.

fingolfin commented:
@cdwensley indeed that worked, but the commit message is a bit odd. Is that intentional, otherwise, as I described above, you could change it with `git commit --amend` (followed by another "force push")

cdwensley commented:
Been tied up today but will look at the commit message tomorrow afternoon.

cdwensley commented:
Two attempts made at changing the commit message.

fingolfin commented:
Thanks for all the efforts!

cdwensley commented:
Now I can practice squashing with XMod, etc.  I'm very grateful.
